File: backend/api/batch.py
"""Batch API — CSV parsing and template rendering for batch generation."""

from __future__ import annotations

import csv
import io
import logging
import re
from typing import Any

logger = logging.getLogger(__name__)

_VAR_RE = re.compile(r"\{\{\s*(\w+)\s*\}\}")


class BatchAPI:
    """pywebview js_api mixin for batch operations."""

    def batch_parse_data(self, content: str, **kwargs) -> dict[str, Any]:
        """Parse CSV text into columns and rows.

        Returns ``{status, columns: [...], rows: [{col: val, ...}, ...]}``.
        """
        try:
            content = content.strip()
            if not content:
                return {"status": "error", "message": "No data provided"}

            reader = csv.DictReader(io.StringIO(content))
            columns = reader.fieldnames or []
            rows = [dict(row) for row in reader]

            if not columns:
                return {"status": "error", "message": "No columns detected in CSV"}
            if not rows:
                return {"status": "error", "message": "No data rows found"}

            return {
                "status": "success",
                "columns": list(columns),
                "rows": rows,
            }
        except Exception as e:
            logger.error("Failed to parse batch data: %s", e)
            return {"status": "error", "message": str(e)}

    def batch_render_template(
        self, template: str, rows: list[dict], **kwargs
    ) -> dict[str, Any]:
        """Render ``{{variable}}`` placeholders for each data row.

        Returns ``{status, rendered: [str, ...], errors: [{row, error}, ...]}``.
        """
        try:
            rendered: list[str] = []
            errors: list[dict[str, Any]] = []

            for i, row in enumerate(rows):
                try:
                    result = _VAR_RE.sub(
                        lambda m: row.get(m.group(1), m.group(0)), template
                    )
                    rendered.append(result)
                except Exception as e:
                    rendered.append("")
                    errors.append({"row": i, "error": str(e)})

            return {
                "status": "success",
                "rendered": rendered,
                "errors": errors,
            }
        except Exception as e:
            logger.error("Failed to render batch template: %s", e)
            return {"status": "error", "message": str(e)}

File: src/main/services/batch.ts
/**
 * Batch operations: CSV parsing and template rendering.
 * Ported from src/bun/services/batch.ts — no changes needed (pure logic).
 */

const VAR_RE = /\{\{\s*(\w+)\s*\}\}/g;

interface BatchParseResult {
	status: "success" | "error";
	message?: string;
	columns?: string[];
	rows?: Record<string, string>[];
}

interface BatchRenderResult {
	status: "success" | "error";
	message?: string;
	rendered?: string[];
	errors?: Array<{ row: number; error: string }>;
}

export function batchParseData(content: string): BatchParseResult {
	const trimmed = content.trim();
	if (!trimmed) {
		return { status: "error", message: "No data provided" };
	}

	const lines = trimmed.split("\n").map((l) => l.trim());
	if (lines.length < 2) {
		return { status: "error", message: "No data rows found" };
	}

	// Parse CSV header
	const columns = parseCSVLine(lines[0]!);
	if (columns.length === 0) {
		return { status: "error", message: "No columns detected in CSV" };
	}

	// Parse data rows
	const rows: Record<string, string>[] = [];
	for (let i = 1; i < lines.length; i++) {
		const line = lines[i]!;
		if (!line) continue;
		const values = parseCSVLine(line);
		const row: Record<string, string> = {};
		for (let j = 0; j < columns.length; j++) {
			row[columns[j]!] = values[j] ?? "";
		}
		rows.push(row);
	}

	if (rows.length === 0) {
		return { status: "error", message: "No data rows found" };
	}

	return { status: "success", columns, rows };
}

export function batchRenderTemplate(
	template: string,
	rows: Record<string, string>[],
): BatchRenderResult {
	const rendered: string[] = [];
	const errors: Array<{ row: number; error: string }> = [];

	for (let i = 0; i < rows.length; i++) {
		try {
			const row = rows[i]!;
			const result = template.replace(
				VAR_RE,
				(_match, varName: string) => row[varName] ?? `{{${varName}}}`,
			);
			rendered.push(result);
		} catch (err) {
			rendered.push("");
			errors.push({ row: i, error: String(err) });
		}
	}

	return { status: "success", rendered, errors };
}

function parseCSVLine(line: string): string[] {
	const result: string[] = [];
	let current = "";
	let inQuotes = false;

	for (let i = 0; i < line.length; i++) {
		const char = line[i]!;

		if (inQuotes) {
			if (char === '"') {
				if (i + 1 < line.length && line[i + 1] === '"') {
					current += '"';
					i++;
				} else {
					inQuotes = false;
				}
			} else {
				current += char;
			}
		} else {
			if (char === '"') {
				inQuotes = true;
			} else if (char === ",") {
				result.push(current.trim());
				current = "";
			} else {
				current += char;
			}
		}
	}

	result.push(current.trim());
	return result;
}
